### PR TITLE
Fix pytorch trainer

### PR DESCRIPTION
**Pull request by SKRohit**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the **CONTRIBUTING.md** document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have verified the code in my local setup.

### CODE REVIEW

**Comment by SKRohit** [comment]
Fix issue #90

File: zenml/steps/trainer/utils.py [hunk 1]
```diff
@@ -64,7 +64,7 @@ def combine_batch_results(x):
     for batch in x:
         for feature, values in batch.items():
             if isinstance(values, torch.Tensor):
-                temp = torch.squeeze(values).detach().numpy()
+                temp = torch.squeeze(values).detach().cpu().numpy()
```
Comment (htahir1) [inline zenml/steps/trainer/utils.py:67]: I'm not sure super sure whether we want to make this explicitly CPU here. Perhaps adding a flag in the trainer itself would make sense?

**Review by htahir1 — CHANGES_REQUESTED** [review]
Thank you very much for the contribution! I like the idea but perhaps we can make it more an explicitly defined behavior in the trainer constructor rather than an explicit behavior. That way people with GPUs and CPUs both could use it. What do you think?

File: zenml/steps/trainer/utils.py (hunk 1 above)
Comment (SKRohit) [reply zenml/steps/trainer/utils.py:67]: > Thank you very much for the contribution! I like the idea but perhaps we can make it more an explicitly defined behavior in the trainer constructor rather than an explicit behavior. That way people with GPUs and CPUs both could use it. What do you think?

Agree completely. I am thinking of keeping `device` as an attribute of `FeedForwardTrainer` whose default value should be `torch.device("cpu")` and users have the choice to pass the value of this argument also while initializing trainer. `TorchBaseTrainerStep` has no implementation otherwise I was thinking of keeping `device` attribute there since for every `PyTorch` model `device` needs to be configured.

Comment (htahir1) [reply zenml/steps/trainer/utils.py:67]: Makes sense. As a bonus: Can we infer from the environment whether a GPU is available and then set the default value accordingly?

Comment (SKRohit) [reply zenml/steps/trainer/utils.py:67]: > I'm not sure super sure whether we want to make this explicitly CPU here. Perhaps adding a flag in the trainer itself would make sense?

If a `tensor` is already on the desired device doing `.cpu()` or `.cuda()` has no effect in `PyTorch` so this would work fine without the need of any `device` flag.

Comment (htahir1) [reply zenml/steps/trainer/utils.py:67]: Yes but if someone does have a GPU and wants to use it then they should be able to default to it, no?

Comment (SKRohit) [reply zenml/steps/trainer/utils.py:67]: > Makes sense. As a bonus: Can we infer from the environment whether a GPU is available and then set the default value accordingly?

Yes, we sure we can. Will do.

Comment (SKRohit) [reply zenml/steps/trainer/utils.py:67]: > Yes but if someone does have a GPU and wants to use it then they should be able to default to it, no?

I need a bit more context here @htahir1. So `temp` variable is going to be used to create `results` `dict` object which contains the output of `model` on `test_dataset`. This `result` object is being saved as `TFRecords` file in the current code. I am assuming (not much familiar with `tfma`) `result` object or its equivalent `TFRecods` files will be used in evaluation step. If that is so putting `results` on GPU will just consume VRAM without any computation being performed on them. Is there any other reason to use GPU for this? 

Also, `to_serialized_examples` function itself needs to be changed or another function needs to be added to convert `torch.tensors`  to `floats, ints` or `bytes` before calling `to_serialized_examples`.

Comment (htahir1) [reply zenml/steps/trainer/utils.py:67]: @SKRohit Sorry for the late response. I think @bcdurak knows this part of the code better. @bcdurak would you say we should take this to GPU or stick to CPU?

Comment (SKRohit) [reply zenml/steps/trainer/utils.py:67]: @htahir1 and @bcdurak I have made `device` a `trainer` attribute. If you will let me know I would do the changes in the code to handle `result` in GPU itself.

Comment (htahir1) [reply zenml/steps/trainer/utils.py:67]: `to_serialized_examples` works because we convert to numpy and it can handle that. Your changes now look great except one small detail: Currently ZenML Steps can only accept primitives as input (ints, floats, strings etc) so a `torch.device` won't work. Can we replace the parameter with a boolean (i.e. `gpu` true or false) or a string like `gpu` / `cpu:0`? Otherwise looks great!

**Review by htahir1 — CHANGES_REQUESTED** [review]

File: zenml/steps/trainer/utils.py (hunk 1 above)
Comment (SKRohit) [reply zenml/steps/trainer/utils.py:67]: Yes. But if we use GPU instead of numpy arrays we would have PyTorch tensors and then we would need to change `to_serialize_examples`. So better keep to handle `result` on CPU itself. 

Btw I have implemented what you suggested to change. I have made `device` to be an `str`. I was testing the code and found that somehow the value of `device` provided by user is not used. `self.steps_dict['trainer']['args']` are still being populated with 

`'{"split_mapping": {"train_splits": ["train"], "eval_splits": ["eval"], "test_splits": ["test"]}, "batch_size": 32, "lr": 0.0001, "epochs": 100, "dropout_chance": 0.2, "loss": "binary_crossentropy", "metrics": ["accuracy"], "hidden_layers": [64, 32, 16], "hidden_activation": "relu", "last_activation": "sigmoid", "input_units": 8, "output_units": 1}'` There is no value of `device` even if I passed while adding trainer   `training_pipeline.add_trainer(TorchFeedForwardTrainer(
    loss='binary_crossentropy',
    last_activation='sigmoid',
    output_units=1,
    metrics=['accuracy'],
    device="cuda:0",
    epochs=100))`. 

Could you please let me know why this is happening and where should I look to understand how `self.steps_dict` is being populated?

Comment (hamzamaiot) [reply zenml/steps/trainer/utils.py:67]: I think its because you need to pass the variable to the `super()` function e.g. `super(device=device...)` at the end of the constructor of your step.

Comment (SKRohit) [reply zenml/steps/trainer/utils.py:67]: Yes, passing `device` to `super` worked. I have done the changes. Let me know if it is fine.

File: zenml/steps/trainer/pytorch_trainers/torch_ff_trainer.py [hunk 2]
```diff
@@ -98,7 +101,45 @@ def __init__(self,
             last_activation=self.last_activation,
             input_units=self.input_units,
             output_units=self.output_units,
+            device = device,
             **kwargs)
+    
+    @staticmethod
+    def _assign_device(device:str):
```
Comment (hamzamaiot) [inline zenml/steps/trainer/pytorch_trainers/torch_ff_trainer.py:108]: Perhaps refactor this into `zenml/steps/trainer/pytorch_trainers/utils.py`?

File: zenml/steps/trainer/pytorch_trainers/torch_ff_trainer.py [hunk 3]
```diff
@@ -98,7 +101,45 @@ def __init__(self,
...
+
+        def _choose_default_device():
```
Comment (hamzamaiot) [inline zenml/steps/trainer/pytorch_trainers/torch_ff_trainer.py:110]: If `_assign_device` is  refactored then this can also be taken out as an inner function

File: zenml/steps/trainer/pytorch_trainers/torch_ff_trainer.py [hunk 4]
```diff
@@ -98,7 +101,45 @@ def __init__(self,
...
+            else:
+                dev = torch.device("cpu")
+                print(f"\n*****Using CPU*****\n")
+            return dev           
+        
+
+        if isinstance(device,str):
+            device = device.lower()
+            try:
+                #flexibility to use desired gpu device e.g. device = cuda:1
```
Comment (hamzamaiot) [inline zenml/steps/trainer/pytorch_trainers/torch_ff_trainer.py:124]: Space after `#` to conform to PEP8

**Review by hamzamaiot — CHANGES_REQUESTED** [review]
Almost there! Thank you for your patience here, the proposed change should be minor

**Review by SKRohit — COMMENTED** [review]
sure. will do and push the changes.

**Comment by SKRohit** [comment]
@hamzamaiot I have pushed the changes. Let me know if any other changes required.
